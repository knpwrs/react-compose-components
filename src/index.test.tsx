import React, { ReactChildren } from 'react';
import { render } from 'react-testing-library';
import Compose from './index';

interface Props {
  children: ReactChildren;
}

const Div = ({ children, ...rest }: Props) => <div {...rest}>{children}</div>;
const Li = ({ children }: Props) => <li>{children}</li>;

it('should compose components and attach props with children', () => {
  const { container } = render((
    <Compose
      components={[
        [Div, { className: 'foo' }],
        'ul',
        Li,
      ]}
    >
      Bar
    </Compose>
  ));
  // Explitly manual testing
  const div = container.children[0];
  expect(div.tagName).toBe('DIV');
  expect(div.className).toBe('foo');
  expect(div.children).toHaveLength(1);
  const ul = div.children[0];
  expect(ul.tagName).toBe('UL');
  expect(ul.className).toBe('');
  expect(ul.children).toHaveLength(1);
  const li = ul.children[0];
  expect(li.tagName).toBe('LI');
  expect(li.className).toBe('');
  expect(li.children).toHaveLength(0);
  expect(li.textContent).toBe('Bar');
  // Extra safety
  expect(div).toMatchSnapshot();
});

it('should compose components without children', () => {
  const { container } = render((
    <Compose
      components={[
        ['div', { className: 'foo' }],
        ['i', { className: 'icon-bar' }],
      ]}
    />
  ));
  // Explicitly manual testing
  const div = container.children[0];
  expect(div.tagName).toBe('DIV');
  expect(div.className).toBe('foo');
  expect(div.children).toHaveLength(1);
  const i = div.children[0];
  expect(i.tagName).toBe('I');
  expect(i.className).toBe('icon-bar');
  expect(i.children).toHaveLength(0);
  // Extra safety
  expect(div).toMatchSnapshot();
});
