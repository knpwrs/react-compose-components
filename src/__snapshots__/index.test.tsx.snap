// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should compose components and attach props with children 1`] = `
<div
  class="foo"
>
  <ul>
    <li>
      Bar
    </li>
  </ul>
</div>
`;

exports[`should compose components without children 1`] = `
<div
  class="foo"
>
  <i
    class="icon-bar"
  />
</div>
`;
